packages/react/src/ResourceArrayInput/ResourceArrayInput.tsx

```diff
@@ -1,48 +1,276 @@
+    } else {
+      typedValue = { type: slice.type[0].code, value };
+    }
+
+    if (
+      discriminators.every((discriminator) => {
+        let nestedProp: TypedValue | TypedValue[] | undefined;
+        let elements: InternalTypeSchema['elements'] = slice.elements;
+
+        if (!isEmpty(slice.elements)) {
+          const ed = getElementDefinitionFromElements(slice.elements, discriminator.path);
+          if (ed) {
+            nestedProp = getTypedPropertyValueWithSchema(typedValue.value, discriminator.path, ed);
+          }
+        }
+
+        if (!nestedProp && slice.typeSchema && !isEmpty(slice.typeSchema?.elements)) {
+          elements = slice.typeSchema.elements;
+          const ed = getElementDefinitionFromElements(slice.typeSchema.elements, discriminator.path);
+          if (ed) {
+            nestedProp = getTypedPropertyValueWithSchema(typedValue.value, discriminator.path, ed);
+          }
+        }
+
+        if (!nestedProp) {
+          console.log('WARN Could not determine slice name');
+          return undefined;
+        }
+
+        return arrayify(nestedProp)?.some((v) => matchDiscriminant(v, discriminator, slice, elements));
+      })
+    ) {
+      return slice.name;
+    }
+  }
+  return undefined;
+}
+
+function assignValuesIntoSlices(
+  values: any[],
+  slices: SupportedSliceDefinition[],
+  slicing: SlicingRules | undefined
+): any[][] {
+  if (!slicing || slicing.slices.length === 0) {
+    return [values];
+  }
+
+  // store values in an array of arrays: one for each slice plus another for non-sliced values
+  const slicedValues: any[][] = new Array(slices.length + 1);
+  for (let i = 0; i < slicedValues.length; i++) {
+    slicedValues[i] = [];
+  }
+
+  for (const value of values) {
+    const sliceName = getValueSliceName(value, slices, slicing.discriminator);
+    let sliceIndex = sliceName ? slices.findIndex((slice) => slice.name === sliceName) : -1;
+    if (sliceIndex === -1) {
+      sliceIndex = slices.length; // values not matched to a slice go in the last entry for non-slice
+    }
+    slicedValues[sliceIndex].push(value);
+  }
+
+  // for slices without existing values, add a placeholder empty value
+  for (let i = 0; i < slices.length; i++) {
+    if (slicedValues[i].length === 0) {
+      slicedValues[i].push(undefined);
+    }
+  }
+
+  return slicedValues;
+}
+
 export function ResourceArrayInput(props: ResourceArrayInputProps): JSX.Element {
-  const [values, setValues] = useState(
-    props.defaultValue && Array.isArray(props.defaultValue) ? props.defaultValue : []
-  );
+  const { property } = props;
+  const medplum = useMedplum();
+  const [loading, setLoading] = useState(true);
+  const [slices, setSlices] = useState<SupportedSliceDefinition[]>();
+  // props.defaultValue should NOT be used after this; prefer the defaultValue state
+  const [defaultValue] = useState<any[]>(() => (Array.isArray(props.defaultValue) ? props.defaultValue : []));
```

Comment:
Should the `else` case be `[props.defaultValue]`?

I think we have an `arrayify` util somewhere in core.

Reply:
I wasn't sure about that; I wanted to keep the handling of `props.defaultValue` the same that it was before, which was:

```javascript
const [values, setValues] = useState(
    props.defaultValue && Array.isArray(props.defaultValue) ? props.defaultValue : []
  );
```

The change you're suggesting doesn't feel quite right since it'd imply you could pass a single, non-array value to a component that is array-based and it would "just work". I'd like to have that be more easily enforced by the type system, type-narrowing, etc., but that'll be it's own project since the usage of `any` is so prevalent right now.

Reply:
To be more explicit, it'd be great to stop using `any` throughout the repo (or just within `ResourceForm` to start) since it makes it difficult to impossible to use various features of typescript.